### ISSUE TITLE
raft: add unit tests for `group0_state_machine::merger` (e.g. using the correct `state_id`)

### ISSUE THREAD

**Issue by margdoc** [scylladb/scylladb#14568]
I have found an issue and fixed it in #13962, but it needs testing.

If `group0_state_machine` applies all commands individually (without batching), the resulting current `state_id` -- which will be compared with the `prev_state_id` of the next command if it is a guarded command -- equals the maximum of the `next_state_id` of all commands applied up to this point. That's because the current `state_id` is obtained from the history table by taking the row with the largest clustering key.

When `group0_state_machine::apply` is called with a batch of commands, the current `state_id` is loaded from `system.group0_history` to `merger::last_group0_state_id` only once. When a command is merged, its `next_state_id` overwrites `last_group0_state_id`, regardless of their order.

Let's consider the following situation:
The leader sends two unguarded `write_mutations` commands concurrently, with timeuuids T1 and T2, where T1 < T2. Leader waits to apply them and sends guarded `topology_change` with `prev_state_id` equal T2. 
Suppose that the command with timeuuid T2 is committed first, and these commands are small enough that all of `write_mutations` could be merged into one command.
Some followers can get all of these three commands before its `fsm` polls them. In this situation, `group0_state_machine::apply` is called with all three of them and `merger` will merge both `write_mutations` into one command. After that, `merger::last_group0_state_id` will be equal to T1 (this command was committed as the second one). When it processes the `topology_change` command, it will compare its `prev_state_id` and `merger::last_group0_state_id`, resulting in making this command a no-op (which wouldn't happen if the commands were applied individually).
Such a scenario results in inconsistent results: one replica applies `topology_change`, but another makes it a no-op.

**Comment by kbr-scylla** [comment]
We could write a similar test as @gleb-cloudius did in group0_cmd_merge_test.cc
-- inject a sleep to poll_output so it returns all commands in a single batch, and send 3 commands -- two unguarded with higher state_id first, then one guarded which depends on the higher state_id.

Then send similar 3 commands but individually.

The results should be the same.

Alternatively (or additionally), a more ambitious move would be to extract merger to its own header so we can unit-test it better.

**Comment by kostja** [comment]
I don't think functional tests are appropriate for this problem. If the merger had unit tests, this could have been tested.

**Comment by gleb-cloudius** [comment]
Why non guarded mutations have state_ids at all? As we can see here they can be applied in different order anyway?

**Comment by kbr-scylla** [comment]
It is possible to concurrently run unguarded and guarded commands.

Consider "pure write" requests running concurrently with "read-modify-write" requests, both implemented in terms of mutation commands. The commands from "pure write" requests can be sent as unguarded mutation commands and pipelined effortlessly. But "read-modify-write" requests need to be guarded, whoever sends the command is using a state that they read before creating the mutation, and they need to ensure that when the command is applied, the state is still the same, so we need to compare prev_state_id with current state_id, and the "pure write" commands need to update the state_id.

**Comment by kbr-scylla** [comment]
I don't understand the scenario or the problem.
T2 < T1

T2 is applied first
Then RMW observes T2
Then T1 is applied
Then RMW is applied but since T1 > T2, the current state id is T1 so it's turned into a no-op (RMW expects T2) -- good

**Comment by kbr-scylla** [comment]
> Then "read-modify-write" command can enter the picture and take T2 state as "last"

No, it will take T1 as last because T1 > T2.
That was the bug that @margdoc fixed in https://github.com/scylladb/scylladb/pull/13962/commits/d2a4079bbeef21d051d07e317d871ad54efd1896 and now added a test for

**Comment by gleb-cloudius** [comment]
> I don't understand the scenario or the problem. T2 < T1

T2 should be newer, not older.

> 
> T2 is applied first Then RMW observes T2 Then T1 is applied Then RMW is applied but since T1 > T2, the current state id is T1 so it's turned into a no-op (RMW expects T2) -- good

T2 is applier first RMW observes T2  then T1 is applied, then RMW is applied successfully since the sate id is T2 - bad.

**Comment by kbr-scylla** [comment]
Ok, I see now.

So in general it's not safe to mix unguarded and guarded commands like this, when the guarded commands could depend on the unguarded ones...

Still, state IDs for unguarded they give one thing that can be useful for debugging: history table entries (with descriptions).

**Comment by gleb-cloudius** [comment]
> Still, state IDs for unguarded they give one thing that can be useful for debugging: history table entries (with descriptions).

But not the order of the application...
